[@unbox ] attributes for externals

diml opened the pull request:
This patch aims to generalize the `"float"` annotations on primitives, which means that a C function takes all its argument as unboxed floats and return its result as an unboxed float.

To do that we allow the attribute `[@unbox]` on individual arguments or on the result of an external. How a value is unboxed/boxed is guessed from the type. If the type is abstract one can set what unboxing/boxing to use by specifying a type as payload: `[@unbox: float]`.

This patch allows to not force boxing when using float primitives that are not of the form `(float ->)* float`. Such as `compare_float` , `Pervasives.classify_float` or `Int64.float_of_bits` which can now be defined as follow:

``` ocaml
external compare_float : (float [@unbox]) -> (float [@unbox]) -> int = ...
external classify_float : (float [@unbox]) -> int = ...
external float_of_bits : (int64 [@unbox]) -> (float [@unbox]) = ...
```

`[@unbox]` is supported for `float`, `int32`, `int64` and `nativeint`. For `int64` on i386 the patch assumes that the result will be in `eax:edx`. This is the only part of the patch I am not sure is portable.

One thing I haven't done yet is update the printing: at the moment if `[@unbox]` attributes are not the same between the implementation and interface the compiler will complain but will display the same type twice.

jdimino-js committed:
Support [@unboxed] and [@untagged] attributes

Adding [@unboxed] (resp [@untagged]) on a primitive argument means
that the argument must passed unboxed (resp untagged) to the external
function. Adding [@unboxed] (resp [@untagged]) on the result means
that the external function returns its result unboxed (resp untagged).

The unboxing (resp untagging) method is derived from the type.

Currently unboxing is suported for: float, int32, int64 and nativeint.
Untagging is supported for int.

This patch also increases the cm{i,o,a,x,xa} magic numbers as the type
Primitive.description is changed.

xavierleroy commented:
This is a very good idea.  However, there is indeed a problem with int64 unboxed arguments on 32-bit platforms.  For returning a 64-bit integer from a C function, the calling conventions are pretty simple: use the first two integer registers (eg eax:edx for i386, r0:r1 for ARM, and r3:r4 for PowerPC).  There might be endianness issues, though.  For passing a 64-bit integer to a C function, it's not good enough to pass it like two consecutive 32-bit integer arguments.  For example, in ARM, a C function taking an int32 and then an int64 expects them in r0 and r2:r3, respectively, not r0 and r1:r2.  So, we need to think more about how to handle this in the Cmmgen pass and how to extend Proc.loc_external_{arguments,results} as needed.

mshinwell commented:
Thanks for the comments.  We're working on the register assignment / cmmgen part.

mshinwell committed:
Split 64-bit integers on 32-bit architectures

mshinwell commented:
Well, we did some thinking about how to deal with the splitting of 64-bit integers on 32-bit targets.  In the end we settled upon doing this in Cmmgen, producing Ctuple nodes, with an target endianness-independent semantics given to the resulting nodes.  In proc.mli we then generalized the loc_external_arguments interface such that it accepts and returns [Reg.t array array].  A member with length > 1 of such an array indicates the presence of an argument that must span multiple registers.  We established a convention, for the return value of loc_external arguments, as to how the ordering of such a subarray maps to the various parts of the argument.  For example, for a 64-bit argument being split into two, the lowest-indexed element of such a subarray corresponds to the register that will hold the least significant half of the argument.  Everything then lines up by magic in [Selectgen.emit_extcall_args] as mentioned in a comment.  We provide a separate implementation for the i386 target, since it already uses different logic for stack slot allocation.

The nice thing about splitting at the Cmmgen stage is that we get some further optimization for free; for example, sequences of calls with int64 values threaded through do not involve intermediate (un)boxing operations.  (We did try doing the int64 splitting at the instruction selection phase, but it proved awkward, and did not have this advantage.)

This appears to be working properly on i386 and ARM hard float hardware.  We're next going to work on ppc32 and sparc32, which should reveal any endianness-related bugs.  We have plans for a comprehensive test suite.  We also need to check it works on ARM soft float, where there is already a trick in the compiler relating to float arguments being split.

There are a couple of other proposals:
1. We'd like this to work at type [int] as well---removing and adding the OCaml tag bit---which permits significantly more C functions to be directly called.  (It is possible to write a "C integer" type now, using [int_as_pointer] in certain cases, which permits direct calls and returns but it would be lovely to avoid it.)  Implementing this should require only a small change.  There is a design question as to the semantics of integers being returned to OCaml at type [int]---there could be a truncation at the top bit---but our current sense is that just documenting such behaviour should suffice.  It seems consistent with the behaviour of the FFI macros.  A user is always free to use another, wider integer type if the OCaml integer size is not wide enough.
2. Given point 1, it seems maybe more appropriate to call the attribute "@native", in the sense that it makes arguments use the native procedure call convention.

Any feedback on the work thus far would be appreciated.

lpw25 commented:
> 1. Given point 1, it seems maybe more appropriate to call the attribute "@native", in the sense that it makes arguments use the native procedure call convention.

Another option would be to use `[@unbox]` for the types which are unboxed, and `[@untag]` for the types (`int`) which are untagged. This has the benefit of being clearer about what the optimised representation actually is.

mshinwell commented:
lpw25: Yeah, I wondered about that too, and it seems fine.

ghost commented:
Also I think we should use `[@unboxed]` instead of `[@unbox]` (and `[@untagged]` for `int`). `[@unbox]` makes me think of the action of unboxing which doesn't make sense for the result type.

alainfrisch commented:
Pervasives.classify_float would also benefit from it.

mshinwell commented on asmcomp/arm/selection.ml:
```diff
@@ -89,6 +89,8 @@ inherit Selectgen.selector_generic as super
   Reg.createv (if !fpu = Soft then begin
                  (* Expand floats into pairs of integer registers *)
+                 (* CR mshinwell: we need to check this in conjunction with
+                    the unboxed external functionality *)
```
At least one (potential) problem. :)

mshinwell commented:
We need to fix the naming, write a testsuite, and test on some more exotic hardware.  We should be able to get this all sorted fairly shortly.

mshinwell commented:
Oh, and I would like Xavier to pass judgement once again...

xavierleroy commented:
I had a quick look.  This looks fine and I'm in favor of merging in trunk.  As mentioned above and earlier today, an "untagged" attribute for type "int" would be nice too.

ghost commented:
OK, I'll add `[@@untagged ]` and some tests and I'll merge the patch.

jdimino-js committed:
Refactor Cmmgen.subst_boxed_number

To make it easier to deal with special cases (int64 on 32 bit target).

jdimino-js committed:
Allow tests to be native code only

jdimino-js committed:
Add tests for [@unboxed] and [@untagged]

ghost commented:
I finished the patch and tests. I'll merge it later this week

gasche commented on typing/typedecl.ml:
```diff
@@ -1714,6 +1810,25 @@ let report_error ppf = function
+      fprintf ppf "Don't know how to untag this type. Only int \
+                   can be untagged"
+  | Type_does_not_match_native_repr_attribute_payload kind ->
+      fprintf ppf "Type and payload of [@%s: ] have different %s \
```
the `:` in `[@%s: ]` is a typo, isn't it? (Is there a reason for leaving spaces before all closing brackets in those error messages?)

diml replied:
I have seen attributes written this way several times so I put the spaces here intentionally. I don't mind removing them, do you think it is better without?

diml replied:
For the colon it's because this error is only in the case the attribute is used as `[@unboxed: <type>]` (instead of `[@unboxed]`).

gasche replied:
I'm fine with the spaces, I didn't know it was standard.

I now understand the error message better, but it's not clear to me that referring explicitly to the "payload" (which is a language used to describe the notation used, not its meaning) is helpful here. I know this kind of comment is more helpful when an alternative is proposed, so here is an attempt:

```
"Type %s cannot be %s as a %s" core_type method payload_type
"Types %s and %s may not have the same %s representation." core_type payload_type method
```

In fact the code is manipulating two different kind of objects, "types" and "native representations", and it's unclear to me whether the payload argument is supposed to be any type (like, will using `[@untagged: size]` when `size` is an alias of `int` be advertised, or is supporting this only a choice of the current implementation?) or one of those predefined names that also denote native representations. If the payload is understood as any type, the second proposed message is better, otherwise the first is ok as well.

diml replied:
In case a type is abstract, the payload is supposed to be the real type. So it can be any type. But maybe allowing the user to put a type here is a bad idea after all. I'll remove this part for now

gasche commented on typing/typedecl.ml:
```diff
@@ -1351,6 +1361,85 @@ let transl_exception env sext =
+    with
+    | _, None ->
+      raise (Error (core_type'.ptyp_loc, Cannot_unbox_or_untag_type kind))
+    | None, Some repr -> repr
```
If I understand the code correctly, this case will allow `(abstract_type [@float])`, and assume that the native representation is float, without checking that `abstract_type` can be represented as a float -- whereas the case below checks that the two native representations coincide.

Is this check done elsewhere, or do we intend to trust the programmer annotation?

diml replied:
> Is this check done elsewhere, or do we intend to trust the programmer annotation?

We have to trust the programmer annotation. But I can't remember why I wanted this and this a dubious feature indeed so I'll remove it

gasche commented:
(I understand the PR is ready to merge and wouldn't want my curiosity to delay it. Feel free to merge, we can always make slightly clarifications later if needed.)

I understand that your backend tests are very thorough, but one thing that feels missing, which was very well done in #188 , is testing for the type-checking aspect of the PR. I think it could help to have in the testsuite a few examples of (1) the end-user syntax (with and without a payload for example) and (2) what is expected to type-check and not type-check.

ghost commented:
I'll add some tests for the type-checking aspect

jdimino-js committed:
Update printing of externals with [@unboxed] or [@untagged] attributes

jdimino-js committed:
Add typing tests for [@unboxed] and [@untagged]

gasche commented on asmcomp/cmmgen.ml:
```diff
@@ -1307,21 +1350,44 @@ let rec is_unboxed_number e =
+           ofs = box_offset boxed_number ->
         Cvar unboxed_id
     | Cop(op, argv) -> Cop(op, List.map subst argv)
+    (* 64 bits integer on 32 bit target *)
```
The case below is slightly fishy. Do you think it could be possible to have a testsuite ensuring that this optimization works as expected? I'm thinking of finding a source file that fires this optimization, and comparing the `-dcmm` output to a known reference. In `testsuite/tests/translprim` I have added tests that compare against a given `-dlambda` output, and erase the identifier numbers for robustness (the logic is in `testsuite/makefiles/Makefile.dlambda`), this should be easily adapted for `-c -dcmm` testing.

diml replied:
If we compare the output of `-dcmm` we'll need one reference file for 32 bits and one for 64 bits. It might to be a bit annoying to update these files when the cmm changes.

I think it's enough to write a small programs that fires this optimization and check that nothing was allocated.

gasche replied:
That's a good idea; there is already such a test in `tests/float-unboxing/float_subst_boxed_number.ml`.

diml replied:
I added a similar test in testsuite/tests/int64-unboxing

jdimino-js committed:
Add a test for int64 unboxing

Especially, test that unboxing of int64 works correctly on 32 bit
platforms.

jdimino-js committed:
Allow unboxing of int64 constants on 32 bit targets

jdimino-js committed:
Support passing unboxed int64 between OCaml and C on ppc32

jdimino-js committed:
Update callilng convention on ppc64

On ppc64 (v1 & v2) passing a float argument to a C function must
reserve a normal register.

jdimino-js committed:
Update sparc backend for unboxed primitive arguments/results

jdimino-js committed:
Support Val in the sparc/emit.mlp

Not tested, but without it just can't compile anything

jdimino-js committed:
Update changelog for GPR#167

ghost commented:
Merged in trunk, revisions 16382 to 16395.

The patch has been tested on i386, amd64, arm, arm64, ppc, ppc64 and ppc64le. The only backend that wasn't tested is sparc, I just tested that ocamlopt builds with ARCH set to sparc and that it can build things.
